XamarinCommunityToolkit/Effects/EntryFormattedTextEffect.android.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.ComponentModel;
+using Android.Text;
+using Android.Text.Style;
+using Android.Widget;
+using Xamarin.CommunityToolkit.Android.Effects;
+using Xamarin.CommunityToolkit.Effects;
+using Xamarin.Forms;
+using Xamarin.Forms.Platform.Android;
+
+[assembly: ResolutionGroupName(nameof(Xamarin.CommunityToolkit))]
+[assembly: ExportEffect(typeof(EntryFormattedTextEffect), nameof(EntryFormattedTextEffect))]
+
+namespace Xamarin.CommunityToolkit.Android.Effects
+{
+	/// <summary>
+	/// The FormattedText <see cref="PlatformEffect"/> specific
+	/// to <see cref="Entry"/> controls.
+	/// </summary>
+	public class EntryFormattedTextEffect : PlatformEffect, IDisposable
+	{
+		bool isDisposed;
+
+		/// <summary>
+		/// Gets the native <see cref="FormsEditText"/> control.
+		/// </summary>
+		public new FormsEditText Control => (FormsEditText)base.Control;
+
+		/// <summary>
+		/// Gets the Xamarin.Forms <see cref="Entry"/> element.
+		/// </summary>
+		public new Entry Element => (Entry)base.Element;
+
+		/// <inheritdoc />
+		protected override void OnAttached()
+		{
+			UpdateFormattedText();
+			Control.TextChanged += OnTextChanged;
+		}
+
+		/// <inheritdoc />
+		protected override void OnDetached()
+		{
+		}
+
+		/// <inheritdoc />
+		protected override void OnElementPropertyChanged(PropertyChangedEventArgs args)
+		{
+			base.OnElementPropertyChanged(args);
+
+			if (args.PropertyName == FormattedTextEffect.FormattedTextProperty.PropertyName)
+			{
+				UpdateFormattedText();
+			}
+		}
+
+		void OnTextChanged(object sender, global::Android.Text.TextChangedEventArgs args)
+		{
+			var spannable = (SpannableStringBuilder)args.Text;
+			if (spannable == null)
+			{
+				return;
+			}
+
+			var characterStyleClass = Java.Lang.Class.FromType(typeof(CharacterStyle));
+			var findSpans = spannable.GetSpans(args.Start, args.Start + args.AfterCount, characterStyleClass);
+
+			if (HasSpan())
+			{
+				var startPosition = spannable.GetSpanStart(findSpans[0]);
+				var endPosition = spannable.GetSpanEnd(findSpans[0]);
+
+				spannable.SetSpan(findSpans[0], startPosition, endPosition, SpanTypes.ExclusiveExclusive);
+			}
+
+			bool HasSpan() => findSpans != null & findSpans.Length > 0;
+		}
```

Comment:
The rules here are kind of complicated and it still needs a lot of work to get right. Currently this algorithm works really good if you are inserting text in the middle of a `<Span>` or at the end of a `<Span>`. If your spans collide as in there is no spacing between them odd behaviors happen.

I am open to suggestions on this, but I think this is a great start. I am highlighting this part of the code as this is a key item to get right as in the specification we said we wanted to retain the formatting if the user inserts or removes characters from a span

Reply:
I was looking at adding the different span types after submitting this and I believe this part needs to be re-written. My original approach here used the following rules
1. Attempt to get the closest first span
2. Apply that span to the spannable

This solution worked okay for the happy path of just TextColor, but once I added all the other properties I don't think it will work well and will become very complicated. 

I am thinking of creating a new instance level property on this class that will store all the spans and their native objects. Once the text property has updated I can then easily retrieve all spans related to the segment and invoke `spannable.SetSpan()` for each of those spans. I think this new proposed solution will be easier for us to manage and won't have as many problems as the original event handler only solution

Reply:
This algorithm has been completely re-written to use shared code as much as possible. Let's open a new discussion thread